internal/services/applications/application_app_role_resource_msgraph.go

```diff
@@ -0,0 +1,194 @@
+package applications
+
+import (
+	"context"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	grapherrors "github.com/manicminer/hamilton/errors"
+	"github.com/manicminer/hamilton/msgraph"
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	helpers "github.com/hashicorp/terraform-provider-azuread/internal/helpers/msgraph"
+	"github.com/hashicorp/terraform-provider-azuread/internal/services/applications/parse"
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+	"github.com/hashicorp/terraform-provider-azuread/internal/utils"
+)
+
+func applicationAppRoleResourceCreateUpdateMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).Applications.MsClient
+
+	objectId := d.Get("application_object_id").(string)
+
+	var roleId string
+	if v, ok := d.GetOk("role_id"); ok {
+		roleId = v.(string)
+	} else {
+		rid, err := uuid.GenerateUUID()
+		if err != nil {
+			return tf.ErrorDiagF(err, "Generating App Role for application with object ID %q", objectId)
+		}
+		roleId = rid
+	}
+
+	allowedMemberTypesRaw := d.Get("allowed_member_types").(*schema.Set).List()
+	allowedMemberTypes := make([]string, 0, len(allowedMemberTypesRaw))
+	for _, a := range allowedMemberTypesRaw {
+		allowedMemberTypes = append(allowedMemberTypes, a.(string))
+	}
+
+	role := msgraph.AppRole{
+		AllowedMemberTypes: &allowedMemberTypes,
+		ID:                 utils.String(roleId),
+		Description:        utils.String(d.Get("description").(string)),
+		DisplayName:        utils.String(d.Get("display_name").(string)),
+		IsEnabled:          utils.Bool(d.Get("is_enabled").(bool)),
+	}
+
+	if v, ok := d.GetOk("value"); ok {
+		role.Value = utils.String(v.(string))
+	}
+
+	id := parse.NewAppRoleID(objectId, *role.ID)
+
+	tf.LockByName(applicationResourceName, id.ObjectId)
+	defer tf.UnlockByName(applicationResourceName, id.ObjectId)
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			return tf.ErrorDiagPathF(nil, "application_object_id", "Application with object ID %q was not found", id.ObjectId)
+		}
+		return tf.ErrorDiagPathF(err, "application_object_id", "Retrieving Application with object ID %q", id.ObjectId)
+	}
+
+	if d.IsNewResource() {
```

Comment:
In 2.0 I'm planning to make `role_id` mandatory as this is the only way to identify a role (same applies to oauth2_permissions), I'll split up the CreateUpdate func then as it will be a lot easier to support delta updates?